buffer/src/main/java/io/netty/buffer/PoolArena.java

```diff
@@ -253,73 +415,224 @@ void reallocate(PooledByteBuf<T> buf, int newCapacity, boolean freeOldMemory) {
+        return max(numTinyAllocations() - numTinyDeallocations(), 0);
+    }
+
+    @Override
+    public long numActiveSmallAllocations() {
+        return max(numSmallAllocations() - numSmallDeallocations(), 0);
+    }
+
+    @Override
+    public long numActiveNormalAllocations() {
+        final long val;
+        synchronized (this) {
+            val = allocationsNormal - deallocationsNormal;
+        }
+        return max(val, 0);
+    }
+
+    @Override
+    public long numActiveHugeAllocations() {
+        return max(numHugeAllocations() - numHugeDeallocations(), 0);
+    }
+
+    @Override
+    public long numActiveBytes() {
+        long val = activeBytesHuge.value();
+        synchronized (this) {
+            for (int i = 0; i < chunkListMetrics.size(); i++) {
+                for (PoolChunkMetric m: chunkListMetrics.get(i)) {
+                    val += m.chunkSize();
+                }
+            }
+        }
+        return max(0, val);
+    }
+
+    protected abstract PoolChunk<T> newChunk(int pageSize, int maxOrder, int pageShifts, int chunkSize);
+    protected abstract PoolChunk<T> newUnpooledChunk(int capacity);
+    protected abstract PooledByteBuf<T> newByteBuf(int maxCapacity);
+    protected abstract void memoryCopy(T src, int srcOffset, T dst, int dstOffset, int length);
+    protected abstract void destroyChunk(PoolChunk<T> chunk);
+
+    @Override
+    public synchronized String toString() {
+        StringBuilder buf = new StringBuilder()
+            .append("Chunk(s) at 0~25%:")
+            .append(StringUtil.NEWLINE)
+            .append(qInit)
+            .append(StringUtil.NEWLINE)
+            .append("Chunk(s) at 0~50%:")
+            .append(StringUtil.NEWLINE)
+            .append(q000)
+            .append(StringUtil.NEWLINE)
+            .append("Chunk(s) at 25~75%:")
+            .append(StringUtil.NEWLINE)
+            .append(q025)
+            .append(StringUtil.NEWLINE)
+            .append("Chunk(s) at 50~100%:")
+            .append(StringUtil.NEWLINE)
+            .append(q050)
+            .append(StringUtil.NEWLINE)
+            .append("Chunk(s) at 75~100%:")
+            .append(StringUtil.NEWLINE)
+            .append(q075)
+            .append(StringUtil.NEWLINE)
+            .append("Chunk(s) at 100%:")
+            .append(StringUtil.NEWLINE)
+            .append(q100)
+            .append(StringUtil.NEWLINE)
+            .append("tiny subpages:");
+        appendPoolSubPages(buf, tinySubpagePools);
+        buf.append(StringUtil.NEWLINE)
+           .append("small subpages:");
+        appendPoolSubPages(buf, smallSubpagePools);
         buf.append(StringUtil.NEWLINE);
-        buf.append("small subpages:");
-        for (int i = 1; i < smallSubpagePools.length; i ++) {
-            PoolSubpage<T> head = smallSubpagePools[i];
+
+        return buf.toString();
+    }
+
+    private static void appendPoolSubPages(StringBuilder buf, PoolSubpage<?>[] subpages) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Consider using varargs for methods or constructors which take an array the last parameter. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)